plugins/dell-dock/fu-dell-dock-plugin.c

```diff
@@ -118,61 +147,170 @@ fu_dell_dock_plugin_get_ec(GPtrArray *devices)
+		g_autofree const gchar *instance_guid = NULL;
+		g_autoptr(FuDellDockHub) hub_device = NULL;
+		g_autoptr(FuDeviceLocker) locker = NULL;
+
+		instance_guid = fwupd_guid_hash_string(instance_id);
+		if (is_newdev) {
+			hub_device = fu_dell_dock_hub_new(FU_USB_DEVICE(device));
+			locker = fu_device_locker_new(FU_DEVICE(hub_device), error);
+			if (locker == NULL)
+				return FALSE;
+
+			fu_device_add_instance_id(FU_DEVICE(hub_device), instance_id);
+			fu_device_add_guid(FU_DEVICE(hub_device), instance_guid);
+			fu_device_add_child(ec_device, FU_DEVICE(hub_device));
+			fu_plugin_device_add(plugin, FU_DEVICE(hub_device));
+		} else {
+			fu_device_add_instance_id(device, instance_id);
+			fu_device_add_guid(device, instance_guid);
+			fu_device_add_child(ec_device, device);
+			fu_plugin_device_add(plugin, device);
+		}
 	}
 
-	/* determine dock type by ec */
-	dock_type = fu_dell_dock_get_dock_type(ec_device);
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_plugin_ec_add_cached_devices(FuPlugin *plugin, FuDevice *ec_device, GError **error)
+{
+	guint8 dock_type;
+	g_autofree const gchar *instance_id = NULL;
+
+	/* determine dock type said by ec */
+	dock_type = fu_dell_dock_plugin_get_dock_type(ec_device, error);
 	if (dock_type == DOCK_BASE_TYPE_UNKNOWN) {
 		g_set_error(error,
 			    FWUPD_ERROR,
 			    FWUPD_ERROR_READ,
 			    "can't read base dock type from EC");
 		return FALSE;
 	}
-	fu_dell_dock_hub_add_instance(FU_DEVICE(hub), dock_type);
-	fu_plugin_device_add(plugin, FU_DEVICE(hub));
-
-	/* add hub instance id for the cached device */
-	hub_dev = fu_plugin_cache_lookup(plugin, hub_cache_key);
-	if (hub_dev != NULL) {
-		fu_dell_dock_hub_add_instance(FU_DEVICE(hub_dev), dock_type);
-		fu_plugin_device_add(plugin, FU_DEVICE(hub_dev));
-		fu_plugin_cache_remove(plugin, hub_cache_key);
+
+	for (guint i = 0; dock_hwdev_info_hub[i].instance_id != NULL; i++) {
+		FuDevice *device;
+		guint16 vid, pid;
+		const gchar *key;
+
+		if (dock_hwdev_info_hub[i].dock_type != dock_type)
+			continue;
+
+		vid = dock_hwdev_info_hub[i].vid;
+		pid = dock_hwdev_info_hub[i].pid;
+		key = g_strdup_printf("USB\\VID_%04X&PID_%04X", vid, pid);
+
+		device = fu_plugin_cache_lookup(plugin, key);
+		if (device != NULL) {
+			if (!(fu_dell_dock_plugin_device_add(plugin, device, TRUE, error)))
+				return FALSE;
+
+			fu_plugin_cache_remove(plugin, key);
+		}
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_plugin_backend_device_added(FuPlugin *plugin,
+					 FuDevice *device,
+					 FuProgress *progress,
+					 GError **error)
+{
+	g_autofree const gchar *instance_id = NULL;
+	guint16 vid, pid;
+
+	/* not interesting */
+	if (!FU_IS_USB_DEVICE(device))
+		return TRUE;
+
+	/* VID and PID */
+	vid = fu_usb_device_get_vid(FU_USB_DEVICE(device));
+	pid = fu_usb_device_get_pid(FU_USB_DEVICE(device));
+
+	/* USB HUB HID bridge device */
+	if ((vid == DELL_DOCK_HID_VID && pid == DELL_DOCK_HID_PID)) {
+		/* API version 1 */
+		{
```

Comment:
weird indentation.  Is this because of getting ready for the K2 changes to add an "if" block around it and not need to do the ping pong for indentation?

Reply:
Yeah. :) k2 section underway here, the indentation attempts to avoid heavy changes in the next PR.